Support for VTN-to-VEN Reports

gthole opened the pull request:
Allow clients to implement VTN-to-VEN reports with poll callback handlers.

Add callback handlers for `oadrRegisterReport` and `oadrUpdateReport` Poll messages.

Notes:
- `oadrRegisterReport` will still return an `oadrRegisteredReport` with no requested reports as before, so there is no change to that behavior unless a VEN client adds a handler.
- `oadrUpdateReport` previously would crash the client since it would attempt to call an `_on_report` method that does not exist on the Client class. Now it will return an `oadrUpdatedReport` to the VTN and call a handler.

Many thanks for this excellent project!

gthole committed:
Allow handlers for register-report and update-report

Signed-off-by: Greg Thole <gthole@gmail.com>

gthole committed:
Fix callbacks to be at report-level

Signed-off-by: Greg Thole <gthole@gmail.com>

gthole committed:
Fix on_report callback

Signed-off-by: Greg Thole <gthole@gmail.com>

axmsoftware reviewed (approved):
Implements a missing feature

axmsoftware committed:
Merge branch 'main' into vtn-reporting-hooks

axmsoftware reviewed (commented):
Test failures need to be addressed before merge .See test-suite workflow for python 3.10

Binbasri-in commented on openleadr/client.py:
```diff
@@ -194,8 +194,9 @@ def add_handler(self, handler, callback):
-            logger.error("'handler' must be either on_event or on_update_event")
+        valid = ('on_event', 'on_update_event', 'on_register_report', 'on_report_update')
+        if handler not in valid:
+            logger.error("'handler' must be in %s" % (valid,))
```
The test case for this needs to be updated to include the new handlers, something like this:

test/test_client_misc.py::test_wrong_handler_supplied 

```
assert any(
        rec.levelname == "ERROR"
        and rec.message.startswith("'handler' must be in")
        for rec in caplog.records
    )
```

Binbasri-in commented on openleadr/client.py:
```diff
@@ -1141,6 +1160,35 @@ async def _execute_hooks(self, hook_name, *args, **kwargs):
 
+    async def _on_register_report(self, response_payload):
+        report_requests = []
+        for report_metadata in response_payload['reports']:
```
The problem is that in the test case it doesn't have the key 'reports', so is using this would solve the problem:

```
response_payload.get('reports', [])
```

or it has to have the metadata and it is not allowed if it doesn't have 'reports '

I will check more

Binbasri-in committed:
Update test_client_misc.py

Included the new valid handlers in the testing, should consider adding test for each?

Signed-off-by: Mohammed Ali <mohammed.binbasri@gmail.com>

Binbasri-in committed:
Update client.py

considered the case when it is empty and no key for 'reports' is there. I think migrating to pydantic would be better

Signed-off-by: Mohammed Ali <mohammed.binbasri@gmail.com>
